### ISSUE TITLE
Erratic aggroup highlighting behaviour

### ISSUE THREAD

**Issue by amerlyq** [albfan/ag.vim#39]
Go to some dir. Lets `t/fixture`. Call there vim and make search in it `AgGroup s`.
Group line highlight can disappear after second. Then move down -- and it will re-appear.
When we stop on `:3` -- it will disappear again and returns after moving to `:4`.
Moreover, if we move up from there, on `:3-:2` highlight again disappears, but when moving in contrary direction `:1-:2` -- it's not. Xxx?

**Comment by albfan** [comment]
I did realize that. I think is a matter of overlapping highlight matchs.

SyntaxAt function should help to detect underneath problem

**Comment by albfan** [comment]
I suppose there's a complex calculation behind detect first file with a prior empty line.

As aggroup files are not supposed to change (except from remove matches) is easy to "mark" lines where "agFile" should apply

**Comment by amerlyq** [comment]
Truly speaking, you invented square wheel.
The fact your naive regexes don't allow to highlight filenames don't mean that is impossible.
And you doing the same thing I did, when re-splitted all my branch -- you mixed completely different changes.

**Comment by amerlyq** [comment]
Don't merge, please, `aggroup` until we have consensus and I review it. 
Because signatures, used by me, had definite sense and goal for next refactoring, and I would liked to understand if them help or interfere.

And move syntax in separate branch as issue which must be solved separately. 
In your current commit there no way to integrate any alternative solution.

**Comment by amerlyq** [comment]
Alright, I will write syntax file for you in separate branch.

**Comment by albfan** [comment]
When I mean complicate to match I mean complicate for vim

http://vimdoc.sourceforge.net/htmldoc/pattern.html#:match
http://vimdoc.sourceforge.net/htmldoc/options.html#'redrawtime'

I have that working but as seen it is erratic

https://github.com/albfan/ag.vim/blob/master/syntax/ag.vim#L19

This codebase change is a step to avoid raw cmdline, so after some discuss should be the way to go. The only change here affecting rest of plugin is to change calls like:

```
call ag#whatever#whatever(arg1, arg2)
```

where you have 

```
let arg { 'arg1':arg1, 'arg2': arg2}
```

is to translate those calls where they are really needed.

```
call ag#whatever#whatever(arg)

...
function ag#whatever#whatever(arg)
...
let arg1 = a:arg.arg1
...
etc
```

the idea is to keep all stuff avaliable at any time to avoid translate or decorate args and viceversa, like 

https://github.com/albfan/ag.vim/blob/master/autoload/ag/args.vim#L39

which broke highlight here:

https://github.com/albfan/ag.vim/blob/master/autoload/ag/group.vim#L30
https://github.com/albfan/ag.vim/blob/master/syntax/ag.vim#L25

and overcomplicate to know user input.

**Comment by amerlyq** [comment]
I don't understand what you mean by 'complicated for vim' at all.
I have studied `h syntax.txt` completely many times. There no way our usecase will be too complex.
Even if you used 10000 matches to remain buffer highlighted you simply need to configure `:syn sync`.
Wait on my syntax file, lets postpone that part of discussion until that. 
Simply remove your changes related to syntax from your commit, or at least split it to drop commit cleanly when needed.

As for signature changes, I understand only half of your point, but for further discuss I need to review and re-think your changes. So, I will mention it later in corresponding issue.

**Comment by albfan** [comment]
I can find the origin of erratic highlight, so I suppose is only vim stopping to calculate matches for whatever reason. The fact that explicit match on lines works is for me a confirmation, but maybe problem is other.

Another easy way to discuss these nasty changes is to see your work in progress refactor. That way I can better understand what I'm broken to you and adapt these changes.

**Comment by amerlyq** [comment]
Roots of erratic behaviour are the overlapping bounds of matching regions, having minuscule but irritating consequences, therefore I called your syntax as naive.
I have already implemented basic syntax for plain highlighting results, but as I wanted more (like individual syntax highlighting for each group), I continued to try it out in different ways, searching most efficient.

Currently there was detected two great problems, which are somewhat unconquerable in their current representation.
- Each line of matching ag results has `lin-` or `lin:col:` prefix, which takes part in embedded syntax matching (even being concealed), and therefore all syntax groups of that language, which use start-of-line `^` can't be matched at all.
  - Possible mitigation: when read ag results into vim buffer -- split out linenumbers into separate array, and use values from that array, when pressing `<CR>` on matching results. But then we completely lose the ability to see linenumbers, when `conceallevel=0`. However we could insert and draw linenumbers again manually like it was done in `easymotion` or `vim-sneak`. Complexity of such solution isn't nice.
  - Do you have ideas or could investigate more info? I couldn't google out something appropriate to force all embedded syntax to ignore linenumber or treat `^` as starting after linenumber, or skip fixed columns count...

If first one had complex but complete and certainly possible solution, second one is truly solemn:
- We can't detect filetype by its path only. We need to respect modelines, because most of text files on unix have no extension at all. 
  - Loading files from disk to memory alredy was done by kernel, when ag searched their content, so we ignore this overload. 
  - But filetype detection in vim triggered mostly by autocmd on opening/creating file, therefore we can't re-use this mechanics as it involves also triggering millions of other functions beside filetype detection, attached to those autocmds.
  - Making and maintaining custom script to detect filetype is great overhead and code duplication. It will never be complete, repeats vim mechanics but in less efficient vimscript.

Therefore I don't have any untried ideas how to resolve it. Can you propose something to do?
If not,  we can abandon embedded syntax completely.
Alternatively I can make command to enable embedded syntax for one filetype only.
Be sure, it's not the same as `setf ...`.
It can be specified manually `:AgFt ...` or even some heuristics could be attached to it -- like inhereting syntax from currently viewed buffer, or remembering last used filetype.

However, different syntax would be more wished, so if you have ideas on filetype detection, I will try work it out. But first problem with lines prefixes remains in place anyways.

**Comment by amerlyq** [comment]
Alright, currently I did single filetype embedding. It's crafted precisely and re-splitted into nice commits.
Issue closed, therefore multifiletype Qs I will move into separate issue. Write your thoughts on that there.
